C++/C++.sublime-syntax

```diff
@@ -69,6 +71,7 @@ variables:
   other_keywords: 'typedef|nullptr|{{visibility_modifiers}}|static_assert|sizeof|using|typeid|alignof|alignas|namespace|template'
   modifiers: '{{storage_classes}}|{{type_qualifier}}|{{compiler_directive}}|{{constant_expression_specifiers}}'
   non_angle_brackets: '(?=<<|<=)'
+  module_identifier: '(\:?[{{identifier}}.])(\:?[{{identifier}}.])?'
```

Comment:
We probably want to scope punctuation separately and only each `{{identifier}}` as `entity`.

Reply:
You mean get rid of `module_identifier` and instead, on line 611, match on {{identifier}} and push a scope that checks for a partition and scopes that partition separately?

```
push:
  - match: '{{identifier}}'
    scope: meta.module.c++ entity.name.module.c++
    push:
      - match: ':{{identifier}}'
        scope: meta.module.c++ entity.name.module.partition.c++
        pop: true
      - match: ''
        pop: true
  - match: ':private'
    scope: meta.module.c++ entity.name.module.private.c++
```

Reply:
The goal was to achieve something like ...

```C++
export   module   A:B;
/* <- meta.module.c++ keyword.control.module.c++ */
/*       ^^^^^^ keyword.control.module.c++ */
/*                ^ entity.name.module.c++ */
/*                 ^ punctuation.accessor.c++ */
/*                  ^ entity.name.module.c++ */

export   module   A : B;
/* <- meta.module.c++ keyword.control.module.c++ */
/*       ^^^^^^ keyword.control.module.c++ */
/*                ^ entity.name.module.c++ */
/*                  ^ punctuation.accessor.c++ */
/*                    ^ entity.name.module.c++ */
```

Not sure if the second one is valid and needs to be supported?

Is something like `A : B : C`  valid?

If module identifiers are simple enough a pattern with capture groups would do it as well. Whatever works best.

Reply:
Ah.  So, the colon after the module name has two uses.  One, to identify the partition.  Two, the identify the :private module fragment.  You can't have multiple partitions.
```c++
export module my.module.name:subpartition;
module :private;
```
The second test is just testing whitespace between module and the partition.
Do you think something like this would be better?
```c++
export module ABC:partD;
/* <- meta.module.c++ keyword.control.module.c++ */
/*     ^^^^^^ keyword.control.module.c++ */
/*            ^^^ entity.name.module.c++ */
/*               ^ punctuation.accessor.c++ */
/*                ^^^^^ entity.name.modulepartition.c++ */
```
I had chosen to include the colon as part of the partition name.  I can change it so it is instead a `punctuation.accessor.c++` type.

I can also change it so the module name is named `entity.name.module.c++`, the partition is named `entity.name.modulepartition.c++`, and `:private` is named `entity.name.moduleprivatefragment.c++`, which all more closely align with the actual terms used for them.

Also, double checking shows that I either have to keep `{{module_identifier}}` or change the code to instead refer to `[{{identifier}}\.]`.  The module name is an identifier plus any number of period separators.  I would prefer keeping `{{module_identifier}}`.

Reply:
>  I can change it so it is instead a `punctuation.accessor.c++` type.

This would help to separate the module name from the partition visually.

> the partition is named `entity.name.modulepartition.c++`

We need to keep in mind that ST derives symbol kinds from those `entity.name` scopes. Such a custom name wouldn't be recognized and thus wouldn't receive any kind icon in the symbol list.

I'd also suggest to group all `:<identifier>` parts under one scope.

Not sure if it makes sense but, something like ...

- `:partID` = `entity.name.partition.other` 
- `:private` = `entity.name.partition.private`

The goal is to enable color schemes to tint all partitions with one simple rule.

> The module name is an identifier plus any number of period separators.

This is not what `[{{identifier}}\.]` pattern descripes. The breakets turn it into a character class, which matches a single char only, while `identifier` is a pattern meant to match tokens.

Wouldn't it need to be `{{identifier}}\.+` instead?

With `export module my.module.name:subpartition;` we need the "push into context" strategy to properly scope all periods `punctuation.accessor`.

Reply:
I took your suggestions on the names.  Partitions are now `entity.name.partition.other` and the private fragment is now `entity.name.partition.private`.

> With `export module my.module.name:subpartition;` we need the "push into context" strategy to properly scope all periods `punctuation.accessor`.

I did not do this as the periods are actually part of the module name.  There is no special functionality to the periods.  They are allowed in the name as a convenience thing so developers can better establish a visual hierarchy with their modules.

Reply:
I also pushed a new version based on all the conversations so far.